### PR TITLE
Express route subgenerator and test added

### PR DESCRIPTION
**Pull request by UndeadBaneGitHub**

### CODE REVIEW

File: express-route/templates/_.server.routes.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+'use strict';
+
+module.exports = function(app) {
+	// Routing logic   
+	// ...
+};
```
Comment (codydaig) [inline express-route/templates/_.server.routes.js]: New Line

**Comment by codydaig** [comment]
Left a line comment.
Don't forget to squash. 
Otherwise, LGTM.

File: test/express-route.test.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+var path = require('path'),
+  helpers = require('yeoman-generator').test,
+  assert = require('yeoman-generator').assert,
+  temp = require('temp').track();
+
+describe('Express Sub Generators Tests', function () {
+  this.timeout(0);
```
Comment (ilanbiala) [inline test/express-route.test.js]: Why no timeout?

File: test/express-route.test.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    temp.cleanup();
+  });
+
+  describe('Generate an express route through the sub-generator', function () {
+    beforeEach(function (done) {
+      helpers.run(path.join(__dirname, '../express-route'))
+        .withOptions({
+          'skip-install': true
+        })
+        .withArguments(['foo'])
```
Comment (ilanbiala) [inline test/express-route.test.js]: What does this argument do?

File: test/express-route.test.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        .on('ready', function (generator) {
+          // this is called right before `generator.run()` is called
+        })
+        .on('end', function () {
+          done();
+        });
+    });
+
+    it('should generate an express route file', function () {
+      assert.file('modules/core/server/routes/foo.server.routes.js');
```
Comment (ilanbiala) [inline test/express-route.test.js]: Shouldn't this be core?

File: test/express-route.test.js (hunk 2 above)
Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: CI tests failed with timeouts, and the `app.test.js` already had the timeout disabled (probably, because git), so I went in the same direction.

Strictly speaking, timeout testing is not really applicable for the generators anyways - the speed is much more dependent on the file system, than on the actual code itself. So, the timeouts were disabled.

File: test/express-route.test.js (hunk 3 above)
Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: It's the name of the generated object.

File: test/express-route.test.js (hunk 4 above)
Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: Hm. True, there should be one route file per module.
Fixing.

Comment (ilanbiala) [inline test/express-route.test.js]: @codydaig thoughts?

Comment (mleanos) [inline test/express-route.test.js]: IMO, the user should be able to specify a custom name here. 

The typical use case would be for creating an additional route config file; to an already existing module.

Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: @mleanos Well, true, that it can be the case - but I'm slowly moving towards "make route/test/controller etc. name inherited from module name".

Reason is, having multiple sets of logics within one module different enough to have separate routes is a confusing and poor, _very_ poor and confusing design. If someone REALLY wants to hack like that - welcome, MEAN boilerplate supports it. But they would have to do it manually, hacky, and the generator would have the best practicies only in it.

Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: I will be devil's advocate as well:
at the same time, here is the example, where having multiple controllers/tests/models etc. in one model might make sense structure-wise.
Suppose, you have "clothes" module. It would be an ok design to group under it both controller/model/routes for "clothes", but also for subcategories of closes, like "pants", "t-shirts", "socks".

If there are not too many subcategories, then merging all of them under the same folder kinda makes sense.

Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: And, there is a "compromise" variant: name moves to prompts, and description for the name prompt can look like this: "Enter route name (leave the name blank to inherit it from module's name):". More-or-less covers both cases.

Comment (mleanos) [inline test/express-route.test.js]: Quick question.. If the new component (route, in this case) be named if it was *inherited* from the module they are adding it to?

Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: @mleanos Can you rephrase the question? Didn't get it, probably a typo somewhere :)

Comment (mleanos) [inline test/express-route.test.js]: Doesn't it make more sense that when someone is adding a route file to an existing module, that a route file would already exist?

Is it likely that the user would be using these sub-generators to create a module one component at a time?

I've found myself creating the whole server-side of a module first. Then going through and customizing it to my needs. 

What's the most likely use case?

Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: @mleanos Hard to tell, to be honest. There can be a use case, where you start with client side first (e.g. create a mockup) and then add server side to it. I'm quite sure that's not the most common usecase, but still a valid one.

**Comment by UndeadBaneGitHub** [comment]
As an experiment, changed it to the "prompt, it left blank inherit from module" behavior.
@ilanbiala @codydaig Take a look, please - if it's fine, I can do likewise in all the other generators.

File: test/express-route.test.js (hunk 4 above)
Comment (mleanos) [inline test/express-route.test.js]: I'd say to keep things simple, and flexible for the user's needs, we should have a prompt for each component. 

I don't think it's bad design to have multiple routes (components) per module. 

For instance, I have a module for managing static lists that admins can manage. Each list has it's own route file for the CRUD operations.

**Comment by mleanos** [comment]
@UndeadBaneGitHub I'm not sure that we should overwrite an existing file. That could be disastrous when the user accidentally left this prompt blank. 

Can we check to see if a file exists, before overwriting it? What are our options of notifying the user, or handling, when the file exists?

**Comment by UndeadBaneGitHub** [comment]
@mleanos
Yo actually prompts for overwrite if it finds that there is such a file - that's why there's no explicit check

File: test/express-route.test.js (hunk 4 above)
Comment (ilanbiala) [inline test/express-route.test.js]: LGTM, but can we get multiple tests for when we have multiple prompt responses.

Comment (UndeadBaneGitHub) [inline test/express-route.test.js]: @ilanbiala there are actually two tests there - one with name and one without.

Comment (ilanbiala) [inline test/express-route.test.js]: :+1:

Comment (codydaig) [inline test/express-route.test.js]: :+1:

**Comment by codydaig** [comment]
@ilanbiala LGTY?

**Comment by UndeadBaneGitHub** [comment]
@codydaig @ilanbiala I'we performed "last swipe" and fixed the remaining pinpointed issues (`scope -> $scope`). If no more are discovered, I hope the PRs to get merged :) After that I can start working on CRUD + ACL (the trickiest one) generators.

**Comment by ilanbiala** [comment]
LGTM.
